### PR TITLE
[REVIEW] Fix fill() when using string_scalar with an empty string

### LINKED ISSUE #4307: [BUG] fill operation creating null values instead empty strings
When creating a column from empty string scalars, the resulting column contains null strings instead of empty strings. Initially found when testing fill with the Java API and then traced back to the cudf layer.

**Steps/Code to reproduce bug**
In https://github.com/rapidsai/cudf/tree/branch-0.13/cpp/tests/filling/fill_tests.cu -- by modifying the set all test on line 237 (though this issue can be recreated in the single, range, and range with null tests)

[Change line 241](https://github.com/rapidsai/cudf/blob/branch-0.13/cpp/tests/filling/fill_tests.cu#L241)
`auto value = "#" + std::to_string(size * 2);`
to
`auto value = "";`

Build and run tests, results in failed test 78 `78 - FILLING_TEST (SEGFAULT)`
```
/home/ryanlee/cudf/cpp/tests/utilities/column_utilities.cu:44: Failure
78:       Expected: lhs.null_count()
78:       Which is: 100
78: To be equal to: rhs.null_count()
78:       Which is: 0
1/2 Test #78: FILLING_TEST .....................***Exception: SegFault  1.21 sec
```

**Expected behavior**
The test should pass when filling with an empty string -- among other issues the null count should remain 0 because empty string is a valid value.

### PR DESCRIPTION
**Pull request by davidwendt**
Closes #4307 
The `cudf::strings::detail::fill()` function accepts a `string_scalar` for use in assigning to the specified rows in a strings column. A `string_scalar` created with any empty string parameter has no size and so the `_data` member is a nullptr. The `fill()` logic was interpreting the nullptr as a null string. This should use `is_valid()` strictly when checking for null scalar instead. Additionally a _valid_ scalar with a nullptr will needs to be interpreted as an empty string. Tests were added for these cases as well.

Once this was fixed, issues were found using `cudf::test::print` on a strings column with only empty strings. The column-utilities' `to_host()` for strings was interpreting an empty chars child column incorrectly as an all-nulls strings column. Turns out this check is not needed for the all-empty case or all-null case and therefore was removed.

Finally, this fix had a side-effect of requiring an additional fix in the column-utilities-test module that was expecting an empty result for `to_host()` for all-nulls strings column. This assumption is incorrect and the test has been corrected.

### CODE REVIEW

**Comment by davidwendt** [comment]
rerun tests

File: cpp/src/strings/filling/fill.cu [hunk 1]
```diff
@@ -46,10 +46,12 @@ std::unique_ptr<column> fill( strings_column_view const& strings,
...
-    auto execpol = rmm::exec_policy(stream);
-    string_view d_value(nullptr,0);
-    if( value.is_valid() )
-        d_value = string_view(value.data(),value.size());
+    // string_scalar.data() is null for valid, empty strings
+    string_view const d_value = [&] {
+        if( !value.is_valid() )
+            return string_view(nullptr,0);
+        return value.size()==0 ? string_view("",0) : string_view(value.data(),value.size());
+    } ();
```
Comment (karthikeyann) [inline cpp/src/strings/filling/fill.cu]: Why is this a lambda?

Comment (karthikeyann) [inline cpp/src/strings/filling/fill.cu]: Why is this a lambda?

File: cpp/src/strings/filling/fill.cu [hunk 2]
```diff
@@ -61,8 +63,8 @@ std::unique_ptr<column> fill( strings_column_view const& strings,
         [d_strings, begin, end, d_value] __device__ (size_type idx) {
             return ((begin <= idx) && (idx < end)) ? !d_value.is_null() : !d_strings.is_null(idx);
         }, stream, mr );
-    rmm::device_buffer null_mask = valid_mask.first;
     auto null_count = valid_mask.second;
+    rmm::device_buffer null_mask(std::move(valid_mask.first));
```
Comment (karthikeyann) [inline cpp/src/strings/filling/fill.cu]: is reference variable a better option here instead of move?

File: cpp/src/strings/filling/fill.cu (hunk 1 above)
Comment (davidwendt) [reply cpp/src/strings/filling/fill.cu]: This is called an immediately initializing lambda expression (IILE). You can use it to initialize a variable without the initialize-update pattern. Mainly it allows the variable `d_value` to be `const`.

File: cpp/src/strings/filling/fill.cu (hunk 2 above)
Comment (karthikeyann) [reply cpp/src/strings/filling/fill.cu]: because another `std::move(null_mask)` is present at line 99.

Comment (davidwendt) [reply cpp/src/strings/filling/fill.cu]: You are right. That would save an extra move. Code has been updated.

**Pull request merged**